src/connection.cpp

```diff
@@ -350,120 +395,172 @@ namespace oxen::quic
             for (auto it = mid; it != streams.end(); ++it)
             {
                 auto& stream_ptr = it->second;
-                if (stream_ptr and not stream_ptr->sent_fin)
-                    strs.push_back(stream_ptr.get());
+                if (stream_ptr and not stream_ptr->_sent_fin)
+                    channels.push_back(stream_ptr.get());
             }
             for (auto it = streams.begin(); it != mid; ++it)
             {
                 auto& stream_ptr = it->second;
-                if (stream_ptr and not stream_ptr->sent_fin)
-                    strs.push_back(stream_ptr.get());
+                if (stream_ptr and not stream_ptr->_sent_fin)
+                    channels.push_back(stream_ptr.get());
             }
         }
 
+        size_t max_datagram_size = -1;
+
+        // if we have datagrams to send, then we should probably do that
+        if (not datagrams->is_empty())
+        {
+            channels.push_back(datagrams.get());
+            // we only need to make this call if we have datagrams to send
+            // TOFIX: store this value when making the call to send_datagram
+            max_datagram_size = get_max_datagram_size();
+        }
         // This is our non-stream value (i.e. we give stream id -1 to ngtcp2 when we hit this).  We
         // hit it after we exhaust all streams (either they have nothing more to give, or we get
         // congested); it takes care of things like initial handshake packets, acks, and also
         // finishes off any partially-filled packet from any previous streams that didn't form a
         // complete packet.
-        strs.push_back(nullptr);
-        auto streams_end_it = std::prev(strs.end());
+        channels.push_back(pseudo_stream.get());
+        auto streams_end_it = std::prev(channels.end());
 
         ngtcp2_pkt_info pkt_info{};
         auto* buf_pos = reinterpret_cast<uint8_t*>(send_buffer.data());
         pkt_tx_timer_updater pkt_updater{*this, ts};
         size_t stream_packets = 0;
-        while (!strs.empty())
-        {
+        ngtcp2_ssize nwrite = 0;
+        ngtcp2_ssize ndatalen;
+        int datagram_accepted = 0;
+        bool write_successful = false;
```

Comment:
These variables should be inside the loop, just before used.  Otherwise their value can persist across loop iterations.